crates/pessimistic-proof/src/proof.rs

```diff
@@ -1,66 +1,27 @@
-use std::collections::HashMap;
-
 use crate::{
-    bridge_exit::NetworkId,
-    certificate::Certificate,
-    keccak::Digest,
-    local_balance_tree::{merge_balance_trees, BalanceTreeByNetwork},
+    bridge_exit::NetworkId, certificate::Certificate, keccak::Digest,
+    local_state::LocalNetworkState,
 };
 
 /// Represents all errors that can occur while generating the proof.
 #[derive(Debug)]
 pub enum ProofError {
     InvalidLocalExitRoot { got: Digest, expected: Digest },
-    NotEnoughBalance { debtors: Vec<NetworkId> },
+    HasDebt { network: NetworkId },
 }
 
 pub type ExitRoot = Digest;
 pub type BalanceRoot = Digest;
-pub type FullProofOutput = (HashMap<NetworkId, ExitRoot>, HashMap<NetworkId, BalanceRoot>);
-
-/// Returns the updated local balance and exit roots for each network.
-pub fn generate_full_proof(certificates: &[Certificate]) -> Result<FullProofOutput, ProofError> {
-    // Check the validity of the provided exit roots
-    for certificate in certificates {
-        let computed_root = certificate.prev_local_exit_tree.get_root();
-
-        if computed_root != certificate.prev_local_exit_root {
-            return Err(ProofError::InvalidLocalExitRoot {
-                got: computed_root,
-                expected: certificate.prev_local_exit_root,
-            });
-        }
-    }
-
-    // Compute the new exit root
-    let exit_roots: HashMap<NetworkId, ExitRoot> = certificates
-        .iter()
-        .map(|certificate| (certificate.origin_network, certificate.compute_new_exit_root()))
-        .collect();
-
-    // Compute the new balance tree by network
-    let balance_trees: HashMap<NetworkId, BalanceTreeByNetwork> = certificates
-        .iter()
-        .map(|certificate| (certificate.origin_network, certificate.compute_new_balance_tree()))
-        .collect();
-
-    // Merge the balance tree by network
-    let balance_tree_by_network: BalanceTreeByNetwork = merge_balance_trees(&balance_trees);
-
-    // Detect the debtors if any
-    let debtors = balance_tree_by_network
-        .iter()
-        .filter_map(|(network, balance_tree)| balance_tree.has_debt().then(|| *network))
-        .collect::<Vec<_>>();
+pub type LeafProofOutput = (ExitRoot, BalanceRoot);
 
-    if !debtors.is_empty() {
-        return Err(ProofError::NotEnoughBalance { debtors });
-    }
+/// Proves that the given [`Certificate`] can be applied on the given [`LocalNetworkState`].
+pub fn generate_leaf_proof(
+    initial_network_state: LocalNetworkState,
+    certificate: Certificate,
+) -> Result<LeafProofOutput, ProofError> {
+    let mut network_state = initial_network_state;
```

Comment:
This function can be reduced to this:
```suggestion
pub fn generate_leaf_proof(
    mut initial_network_state: LocalNetworkState,
    certificate: Certificate,
) -> Result<LeafProofOutput, ProofError> {
	initial_network_state.apply_certificate(certificate)
}
```

I guess we can't just call the `apply_certificate` directly in the program, right?

Reply:
I prefer to separate like this as for now the statement `generate_leaf_proof` is nothing more than applying the certificate but that will be extended right after